crates/storage/libmdbx-rs/src/transaction.rs

```diff
@@ -560,12 +530,23 @@ impl TransactionPtr {
 
     /// Executes the given closure once the lock on the transaction is acquired.
     #[inline]
-    pub(crate) fn txn_execute<F, T>(&self, f: F) -> T
+    pub(crate) fn txn_execute<F, T>(&self, f: F) -> Result<T>
     where
         F: FnOnce(*mut ffi::MDBX_txn) -> T,
     {
         let _lck = self.lock.lock();
-        (f)(self.txn)
+
+        // When transaction is aborted via `TxnManager`, it's actually reset using `mdbx_txn_reset`
+        // that makes the transaction unusable and sets the `MDBX_TXN_FINISHED` flag.
```

Comment:
what happens inside `mdbx_tnx_reset` is this:
```c
txn->mt_flags = MDBX_TXN_RDONLY | MDBX_TXN_FINISHED;
```
which breaks the following assertion in `check_txn` when running in debug mode:
```c
  tASSERT(txn, (txn->mt_flags & MDBX_NOTLS) ==
                   ((txn->mt_flags & MDBX_TXN_RDONLY)
                        ? txn->mt_env->me_flags & MDBX_NOTLS
                        : 0));
```
and segfaults in release mode

the pointer is valid, because `mdbx_txn_reset` doesn't actually free the pointer, but it becomes unusable as an MDBX transaction and leads to some broken assumptions. I think it's an MDBX issue, and the `check_txn` should instead return an error if `MDBX_TXN_FINISHED` flag is set.

Reply:
upd: MDBX author confirmed it's a bug and not a feature